### ISSUE TITLE
WebSocket not disconnecting when detached

### ISSUE THREAD

**Issue by SillyMoo** [robohydra/robohydra#31]
When a WebSocket head is disconnected, the websocket remains open. This is a problem if you have a long lived websocket and wish to change behaviour by swapping between heads.
It might be good to provide an option to allow disconnection when detached. I am happy to provide a PR for this issue if accepted.

**Comment by emanchado** [comment]
Thanks for the bug report!

Hm, if I understand the code correctly, that's not something you can do automatically because in principle there's no reference to the open socket in the head. Also, there could be several sockets! So it seems to me that the best solution would be to expose a way to run code when the head is detached. At least for `RoboHydraWebSocketHead`s. What did you have in mind? This? Something else?

Are you thinking of redefining the `detach` method for `RoboHydraWebSocketHead` and call some head property (eg. `this.onDetach`) if it was defined on head creation?

**Comment by SillyMoo** [comment]
Sorry for the delay in responding, I got distracted :) So there is a reference to the socket inside the handler function in the head. So the solution I am using at the moment is to add a detach() method from inside the handler, which detaches the socket.
Obviously I had however forgotten that there can be >1 socket (I'm testing a client so only have 1 socket open at a time). But I see no reason you could not keep a list of sockets and close them on detach.
I do however like the idea of an onDetach property/event, makes things more flexible (and presumably the WebSocketProxy would by default detach as you would not usually expect folks to extend the proxy?).

**Comment by emanchado** [comment]
Yeah, that probably makes sense. Are you going to prepare a PR? I'm a bit scared about potentially breaking tests (by changing the default behaviour), but the chances are probably ridiculously low, so I'm guessing we're fine.
